### ISSUE TITLE
question about usage and pre-existing models

### ISSUE THREAD

**Issue by ckot** [charettes/django-mutant#42]
Hi, 

I've looked at the following examples:

[integricho/mutant-sample-app](https://github.com/integricho/mutant-sample-app) 

[charettes/django-mutant-usecase.rst](https://gist.github.com/charettes/5670245)

I have a lot of already defined "static models" in my models.py. Can mutant be used to extend them,  or do they have to be redefined using ModelDefinition.objects.create()?

I thought I could create a ModelDefinition for one of my pre-existing models and it would recognize that it didn't need to create any new database tables and/or content-types, and would simply allow my to add more fields to my model.

Instead I got a UNIQUE constraint error when it tried to insert a duplicate contenttype.

Am I misunderstanding the usage of this package?  Can MutableModel be used as a base class for my models instead of django.models.Model or perhaps add ModelDefinitionAttribute as a mixin?

Any help would be appreciated,

-Scott

**Comment by charettes** [comment]
Hi @ckot,

> I have a lot of already defined "static models" in my models.py. Can mutant be used to extend them, or do they have to be redefined using ModelDefinition.objects.create()?

If you initially create your `ModelDefinition` objects with `managed=False` you should be able reuse your existing model class. You could also define your existing statically defined model classes as `abstract=True` and reuse them `BaseDefinition`.

What I'd like to expose in the future is a way to have both dynamic and static model definition living in two different project state. This way you statically reference your model definition just like any other class and also mark them a mutable (by subclassing `MutableModel` for example) in order to access their dynamic definition on demand using mutan's API (e.g. `definition().model_class()`).

**Comment by ckot** [comment]
Hi @charettes , thanks for your reply,

> 
If you initially create your ModelDefinition objects with managed=False you should be able reuse your existing model class. You could also define your existing statically defined model classes as abstract=True and reuse them with BaseDefinition instances.

I just tried that, but perhaps I didn't understand.  What I did was the following:

**admin.py**

```
for field_type in FieldDefinitionBase._field_definitions.values():
    fld_name = field_type.__name__
    attrs = {'model': field_type}
    FieldDefAdmin = type('{0}Admin'.format(fld_name),
                         (admin.ModelAdmin,),
                         attrs)
    admin.site.register(field_type, FieldDefAdmin)

admin.site.register(ModelDefinition)

```

and then in the admin, I clicked on 'add' ModelDefinition and in the form inserted:

```
App label: <my apps label>
Python Model Class Name: <my static models name>
Object name: <my static models name>
Managed: unchecked
Database table: tried both <app_label_lowercase_model_name> and blank
Verbose name: tried both with a value and blank
Verbose name plural: tried with both a value and blank
```

and I got the same Unique Constraint error.  Were you perhaps saying that I define a model definition for one of my existing models without using the admin?

Oh, and I've always been confused about Python Model Class Name vs Object name.  Are they supposed to be different things?  If manually defining a ModelDefinition it doesn't seem you need to pass a python model class name - just object_name.

-Scott

**Comment by charettes** [comment]
> Were you perhaps saying that I define a model definition for one of my existing models without using the admin?

Yes.

> If manually defining a ModelDefinition it doesn't seem you need to pass a python model class name - just object_name.

`object_name` is the string used to create the class.

**Comment by ckot** [comment]
Sorry if I'm being particularly dense. If you can bear with me, I'll be happy to help with documentation.

Just to make things more concrete lets say I have 2 apps:

in app1.models.py I've defined a static model Experiment

in app2.models.py I've defined a static model Condition

I'd like to dynamically add what would otherwise be done statically by defining the field:
experiment = ForeignKey('Experiment',  related_name='conditions')

to my Condition model in app2.models.py

> object_name is the string used to create the class.

Are you referring to the model definition class or the static Django model class?

I tried the following in manage.py shell

```
exp_def = ModelDefinition.objects.create(app_label='app1', object_name='Experiment', managed=False)
```
and get the Unique Constraint error.  I then instead tried:

```
exp_def = ModelDefinition.objects.create(app_label='app1', object_name='experimentdefinition', managed=False)
```

and I don't get an error, but I don't see how it's associated with 'Experiment'

I tried adding the `model='Experiment'` kwarg, but when I type `vars(exp_def)`, I get:

```
{'_model_class': app1.models.experimentdefinition,
 '_state': <django.db.models.base.ModelState at 0x7fb27f033cd0>,
 'app_label': 'app1',
 'contenttype_ptr_id': 49,
 'db_table': None,
 'id': 49,
 'managed': False,
 'model': 'experimentdefinition',
 'object_name': 'experimentdefinition',
 'verbose_name': None,
 'verbose_name_plural': None}

```

Sorry for being such a pain, but could you provide me with the exact parameters to create a ModelDefinition for my 'Experiment' object.  I should be able to take it from there and the help would be much appreciated.

**Comment by charettes** [comment]
To create the mutable model definition do the following:

```python
ModelDefinition.objects.get_or_create(
    app_label='app1',
    object_name='Experiment',
    defaults={
        'managed': False
    }
)
```

It should reuse the existing `ContentType` associated with your `app1.Experiment` model. The thing is this model won't have any of the fields statically defined on your model.

If you want to create a foreign key from a mutable model use `mutant.contrib.related.ForeignKeyDefinition` (make sure to add `'mutant.contrib.related'` to your `INSTALLED_APPS`).

**Comment by ckot** [comment]
Hmm.  tried that and it still gave the UniqueConstraint error.  

I then tried adding:
```
class Meta:
    abstract = True

```
to my models, blew away my db, and then it worked, but as you said, none of my statically defined fields are defined.  Also, I noticed that I had to specify `db_table='app1_experiment'` so it didn't name the table 'mutant_app1_experiment'

I not opposed to completely redefining my static fields as part of the model definitions, renaming my old models and doing a data migration, but it seems that methods associated with my static models are lost.  Is there a way to attach methods to  the model class associated with a model definition?

**Comment by charettes** [comment]
You don't have to rely to monkey patching, simply use `BaseDefinition`:

```python

class AbstractExperiment(models.Model):
    statically_defined_field = models.TextField()

    class Meta:
        abstract = True

Experiment = ModelDefinition.objects.create(
    app_label='app1',
    object_name='Experiment',
    bases=[BaseDefinition(base=AbstractExperiment)],
)

**Comment by ckot** [comment]
That's the ticket!  Thanks so much, by using BaseDefinition, it even preserves the statically defined fields!

I still need to try this out on static models which use inheritance, but otherwise it's looking good!

**Comment by ckot** [comment]
Hmm, Having some trouble with models which are subclasses of abstract models. Here's what my model actually looks like:

```
class AutoTimeStampModel(models.Model):
    created = models.DateTimeField()
    modified = models.DateTimeField()

    def save(self, *args, **kwargs):
        """On save, update timestamps"""
        if self.id is None:
            self.created = timezone.now()
        self.modified = timezone.now()
        return super(AutoTimeStampModel, self).save(*args, **kwargs)

    class Meta(object):
        abstract = True

    class AbstractExperiment(AutoTimeStampModel):
        statically_defined_field = models.TextField()

        class Meta:
            abstract = True
```

if I do the following:

Experiment = ModelDefinition.objects.create(
    app_label='app1',
    object_name='Experiment',
    bases=[Experiment = ModelDefinition.objects.create(
    app_label='app1',
    object_name='Experiment',
    bases=[BaseDefinition(base=AbstractExperiment),
BaseDefinition(base=AbstractExperiment)],
)

**Comment by ckot** [comment]
Sorry most of my comments above were just myself flailing, I've figured out most of my issues since, but as I only needed to add a single field to a model from one of my reusable modules, [django-swappable-models](https://github.com/wq/django-swappable-models)  turned out a better fit for my project.  I'm still interested in this project though, and think what I've learned through all my struggling will be useful for some documentation.  I'll revisit this once I'm done working on my project.

One question though.  I was simply playing around with defining my modeldefs in `./manage.py shell`. What is the suggested way to do this in actual code.  I played around with doing this in my AppConfig's `ready()`, but I *think* I ran into problems.  I was able to do it in a migration via `RunPy`, but I'm not so crazy about having my models defined in a place where users would be very unlikely to look.
